In [ ]:
#javascript for console of Colab
# function ClickConnect(){
# 	console.log("Working"); 
# 	document.querySelector("colab-toolbar-button").click() 
# } setInterval(ClickConnect, 1800000)

# import Library

In [1]:
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import Dataset
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch.utils.tensorboard import SummaryWriter

from PIL import Image
from collections import OrderedDict #for multiGPU
from tqdm import tqdm

import os
import random
import time

# GPU

In [2]:
# device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
#for colab
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print("device is ", device)

device is  cuda


# Load Data

In [3]:
print("loading dataset")

data_transforms = {
    'train' : transforms.Compose([ #horizontal flip, random crop, normalization           
            #transforms.ToPILImage(),
            # transforms.Resize((256,256)),
            transforms.RandomCrop(32, padding = 4), #augmentation, input for ResNet
            transforms.RandomHorizontalFlip(), #augmentation
            transforms.ToTensor(),
            transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)), #with the per-pixel mean subtracted
            #normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                     #std=[0.229, 0.224, 0.225])

        ]) , 
        'test': transforms.Compose([
            transforms.ToTensor(), #single view of the original 32x32 image
            transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
        ])
    }
#access data_transforms['train']

#dataloader (batch)
train_set = torchvision.datasets.CIFAR10(
    root='./data', train=True, download=True, transform = data_transforms['train']
)
trainloader = torch.utils.data.DataLoader(
    train_set, batch_size = 128, shuffle = True, num_workers=2
)

test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)

#classes
classes = ('plane', 'car', 'bird', 'cat', 'deer','dog', 'frog', 'horse', 'ship', 'truck')


loading dataset


100%|██████████| 170498071/170498071 [00:02<00:00, 70260544.12it/s]


Extracting ./data/cifar-10-python.tar.gz to ./data
Files already downloaded and verified


In [ ]:
print(len(train_set))
print(len(test_set))

50000
10000


# epochs, iterations

In [ ]:
Iterations = 48000
total_epochs = int(Iterations/len(trainloader))

print(total_epochs)

122


# Model Architecture

In [ ]:
# x = torch.zeros(1,16,32,32)
# k = nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False)
# y = k(x)
# print(y.shape)

torch.Size([1, 16, 32, 32])


In [ ]:
# x = torch.zeros(1,16,32,32)
# k = nn.Conv2d(16, 32, kernel_size = 3, stride = 2, padding = 1, bias = False)
# y = k(x)
# print(y.shape)

torch.Size([1, 32, 16, 16])


In [4]:
class optionA(nn.Module):
  def __init__(self, func):
    super(optionA, self).__init__()
    self.func = func
  
  def forward(self, x):
    return self.func(x)

In [5]:
class Fx2_16(nn.Module):
  def __init__(self):
    super().__init__()
    self.conv1 = nn.Sequential(
        nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(16)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(16)
    )
    self.residual = nn.Sequential()
  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out +=self.residual(x)
    out = F.relu(out)

    return out

In [6]:
class Fx2_32(nn.Module):
  def __init__(self):
    super().__init__()
    self.conv1 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32),
    )
    self.residual = nn.Sequential()
  
  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out += self.residual(x)
    out = F.relu(out)
    return out


In [7]:
class Fx2_64(nn.Module):
  def __init__(self):
    super().__init__()

    self.conv1 = nn.Sequential(
         nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.residual = nn.Sequential()

  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out += self.residual(x)
    out = F.relu(out)
    return out

In [8]:
#classes for Resnet20 Model
class conv16 (nn.Module): #Blue
  def __init__(self):
    super().__init__()
    self.Fx1 = Fx2_16()
    self.Fx2 = Fx2_16()
    self.Fx3 = Fx2_16()
    
  def forward(self,x):
    x1 = self.Fx1(x) 
    x2 = self.Fx2(x1)
    x3 = self.Fx3(x2) 
 
    return x3

#classes for Resnet20 Model
class conv32 (nn.Module): #Purple
  def __init__(self):
    super().__init__()
    self.Fx1_1 = nn.Sequential(
        #torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', device=None, dtype=None)
        nn.Conv2d(16, 32, kernel_size = 3, stride = 2, padding = 1, bias = False),
        nn.BatchNorm2d(32))
    self.Fx1_2 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32)
    )
    ##OptionB
    # self.residual_mapping = nn.Sequential(
    #     nn.Conv2d(16,32,kernel_size =1, stride = 2, bias= False),
    #     nn.BatchNorm2d(32)
    # )

    ## OptionA
    self.residual_mapping = optionA(lambda x: F.pad(x[:, :, ::2, ::2], (0, 0, 0, 0, 32//4, 32//4), "constant", 0))

    self.Fx2 = Fx2_32()
    self.Fx3 = Fx2_32()
    
  def forward(self,x):
    out = F.relu(self.Fx1_1(x))
    out = self.Fx1_2(out) 
    out += self.residual_mapping(x)
    x1 = F.relu(out)

    x2 = self.Fx2(x1) 
    x3 = self.Fx3(x2) 

    return x3

#classes for Resnet20 Model
class conv64 (nn.Module): #Green
  def __init__(self):
    super().__init__()
    self.Fx1_1 = nn.Sequential(
        #torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', device=None, dtype=None)
        nn.Conv2d(32, 64, kernel_size = 3, stride = 2, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.Fx1_2 = nn.Sequential(
        nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    ##OptionB
    # self.residual_mapping = nn.Sequential(
    #     nn.Conv2d(32,64,kernel_size =1, stride = 2, bias= False),
    #     nn.BatchNorm2d(64)
    # )

    ## OptionA
    self.residual_mapping = optionA(lambda x: F.pad(x[:, :, ::2, ::2], (0, 0, 0, 0, 64//4, 64//4), "constant", 0))
  
    self.Fx2 = Fx2_64()
    self.Fx3 = Fx2_64()
    
  def forward(self,x):
    out = F.relu(self.Fx1_1(x))
    out = self.Fx1_2(out)
    out += self.residual_mapping(x)
    x1 = F.relu(out)

    x2 = self.Fx2(x1) 
    x3 = self.Fx3(x2) 

    return x3


In [ ]:
# x = torch.zeros(1,16,32,32)
# k = conv16()
# y = k(x)
# print(y.shape)

torch.Size([1, 16, 32, 32])


In [ ]:
# x = torch.zeros(1,16,32,32)
# k = conv32()
# y = k(x)
# print(y.shape)

torch.Size([1, 16, 32, 32])
torch.Size([1, 32, 16, 16])
torch.Size([1, 32, 16, 16])
torch.Size([1, 32, 16, 16])


In [9]:
class MyModel(nn.Module): #ResNet20 architecture implementation
    def __init__(self, conv2 = conv16, conv3 = conv32,
            conv4 = conv64, num_classes = 10, init_weights = True): #num_classes
        super().__init__()
        self.conv1 = nn.Sequential(
            nn.Conv2d(3, 16, kernel_size = 3, stride = 1, padding=1, bias = False),
            nn.BatchNorm2d(16),
            nn.ReLU(),
        )
        self.conv2 = conv2()
        self.conv3 = conv3()
        self.conv4 = conv4()
        #self.avg_pool = nn.AdaptiveAvgPool2d((1,1))
        self.fclayer = nn.Linear(64, num_classes)

        if init_weights:
            self.initialize_weights()
    
    # define weight initialization function
    def initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, 0, 0.01)
                nn.init.constant_(m.bias, 0) 
        
    def forward(self, x):
        x1 = self.conv1(x)
        x2 = self.conv2(x1)
        x3 = self.conv3(x2)
        x4 = self.conv4(x3)
        x = F.avg_pool2d(x4, x4.size()[3])
        #x = self.avg_pool(x4)
        x = x.view(x.size(0), -1)
        output = self.fclayer(x)
        return output

In [ ]:
from torchsummary import summary
resnet20 = MyModel()
summary(resnet20, (3,32,32))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 32, 32]             432
       BatchNorm2d-2           [-1, 16, 32, 32]              32
              ReLU-3           [-1, 16, 32, 32]               0
            Conv2d-4           [-1, 16, 32, 32]           2,304
       BatchNorm2d-5           [-1, 16, 32, 32]              32
            Conv2d-6           [-1, 16, 32, 32]           2,304
       BatchNorm2d-7           [-1, 16, 32, 32]              32
            Fx2_16-8           [-1, 16, 32, 32]               0
            Conv2d-9           [-1, 16, 32, 32]           2,304
      BatchNorm2d-10           [-1, 16, 32, 32]              32
           Conv2d-11           [-1, 16, 32, 32]           2,304
      BatchNorm2d-12           [-1, 16, 32, 32]              32
           Fx2_16-13           [-1, 16, 32, 32]               0
           Conv2d-14           [-1, 16,

In [ ]:
resnet20

MyModel(
  (conv1): Sequential(
    (0): Conv2d(3, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU()
  )
  (conv2): conv16(
    (Fx1): Fx2_16(
      (conv1): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (residual): Sequential()
    )
    (Fx2): Fx2_16(
      (conv1): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Sequent

# Train

In [ ]:
# print("building model")
# net = MyModel()
# net = net.to(device)

building model


In [10]:
# #get_model when loading model parameters trained w/ multi-gpu
def get_model(model_name, checkpoint_path):
    print("start loading model parameter")
    model = model_name()
    state_dict = torch.load(checkpoint_path)

    new_state_dict = OrderedDict()

    for k, v in state_dict.items():
        name = k.replace('module.', '')
        new_state_dict[name] = v
    model.load_state_dict(new_state_dict)
    print("model parameter loaded")
    return model

In [11]:

def train(n_epochs, train_loader) :
 
    epochs = n_epochs
    train_dataloader = train_loader
    #get current time -> folder name
    train_date = time.strftime("%y%m%d_%H%M", time.localtime(time.time()))
    # print(train_date)

    #log path for tensorboard
    log_path = os.path.join(os.getcwd(),"logs",train_date)
    print("log_path is {}".format(log_path))

    #model path to save '.pth'file
    model_path = os.path.join(os.getcwd(),"models",train_date)
    print("model_path is {}".format(model_path))
    if os.path.isdir(model_path) == False:
            os.makedirs(model_path)

    #writer for tensorboard
    writer = SummaryWriter(log_path)
 
    # whether available GPU
    # device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    #device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    #define model (처음부터 학습시)
    print("처음부터 학습")
    model = MyModel()
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model.to(device)

    ####### 이전 학습한거에 추가적으로 학습시
    # print("이전에 하던거 부터 학습")
    # model_name = MyModel
    # # /content/models/230412_2145/0.pth
    # checkpoint_path = './12.pth'  #수정 필요
    # model = get_model(model_name,checkpoint_path)
    # model.to(device)

    #######

    #define loss funcion, optimizer, (scheduler)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(),lr=0.1, momentum=0.9,weight_decay=0.0001)
    #lr = 0.025, momentum = 0.9, weight_decay = 0.0005(original)
    scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [81,122], gamma = 0.1)
    # 0.1 -> 0.01 (32K iterations = 81 epochs) -> 0.001 (48K iterations = 122 epochs)

    print("start training")
    #Train
    for epoch in range(epochs):
        
        model.train(True)
        len_trainbatch = len(train_dataloader)

        i = 0
        for inputs, targets in tqdm(train_dataloader):
            i = i+1
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            

            writer.add_scalar('train/loss', loss.item(), epoch*len_trainbatch+i) #iterations

        scheduler.step()    
        print("train loss of {}th epoch is {}".format(epoch,loss.item())) ## 수정 필요
 
        # Save every N(30) epoch
        save_epoch = 30
        if save_epoch > 0 and epoch % save_epoch == save_epoch-1:
            state_dict = model.state_dict()
            torch.save(state_dict, os.path.join(model_path, str(epoch+1)+'.pth')) #epoch+1로 바꾸기
        
    print('Finished Training')

    #save model after training
    torch.save(model.state_dict(), './model.pth') 

   

# Main

In [12]:

def main() :

    random.seed(53)
    train(163, trainloader) #64k iterations = 163 epochs

   

if __name__ == '__main__':
    main()

log_path is /content/logs/230417_0009
model_path is /content/models/230417_0009
처음부터 학습
start training


100%|██████████| 391/391 [00:36<00:00, 10.73it/s]


train loss of 0th epoch is 1.0763870477676392


100%|██████████| 391/391 [00:27<00:00, 14.32it/s]


train loss of 1th epoch is 0.9113261103630066


100%|██████████| 391/391 [00:27<00:00, 14.42it/s]


train loss of 2th epoch is 0.7935523986816406


100%|██████████| 391/391 [00:27<00:00, 14.17it/s]


train loss of 3th epoch is 0.6975749135017395


100%|██████████| 391/391 [00:27<00:00, 14.23it/s]


train loss of 4th epoch is 0.7071053385734558


100%|██████████| 391/391 [00:27<00:00, 14.28it/s]


train loss of 5th epoch is 0.550421416759491


100%|██████████| 391/391 [00:27<00:00, 14.27it/s]


train loss of 6th epoch is 0.5025944709777832


100%|██████████| 391/391 [00:27<00:00, 14.18it/s]


train loss of 7th epoch is 0.3679409921169281


100%|██████████| 391/391 [00:27<00:00, 14.20it/s]


train loss of 8th epoch is 0.5335637927055359


100%|██████████| 391/391 [00:27<00:00, 14.15it/s]


train loss of 9th epoch is 0.40994420647621155


100%|██████████| 391/391 [00:27<00:00, 14.42it/s]


train loss of 10th epoch is 0.3918271064758301


100%|██████████| 391/391 [00:27<00:00, 14.27it/s]


train loss of 11th epoch is 0.4418517053127289


100%|██████████| 391/391 [00:27<00:00, 14.32it/s]


train loss of 12th epoch is 0.5215147137641907


100%|██████████| 391/391 [00:27<00:00, 14.44it/s]


train loss of 13th epoch is 0.38103044033050537


100%|██████████| 391/391 [00:26<00:00, 14.51it/s]


train loss of 14th epoch is 0.24819283187389374


100%|██████████| 391/391 [00:27<00:00, 14.39it/s]


train loss of 15th epoch is 0.45941418409347534


100%|██████████| 391/391 [00:27<00:00, 14.40it/s]


train loss of 16th epoch is 0.33980268239974976


100%|██████████| 391/391 [00:26<00:00, 14.49it/s]


train loss of 17th epoch is 0.5791303515434265


100%|██████████| 391/391 [00:26<00:00, 14.48it/s]


train loss of 18th epoch is 0.3840877413749695


100%|██████████| 391/391 [00:27<00:00, 14.46it/s]


train loss of 19th epoch is 0.3586638569831848


100%|██████████| 391/391 [00:27<00:00, 14.36it/s]


train loss of 20th epoch is 0.37702131271362305


100%|██████████| 391/391 [00:27<00:00, 14.31it/s]


train loss of 21th epoch is 0.2637298107147217


100%|██████████| 391/391 [00:27<00:00, 14.37it/s]


train loss of 22th epoch is 0.22816577553749084


100%|██████████| 391/391 [00:27<00:00, 14.33it/s]


train loss of 23th epoch is 0.22753708064556122


100%|██████████| 391/391 [00:27<00:00, 14.40it/s]


train loss of 24th epoch is 0.2480258196592331


100%|██████████| 391/391 [00:27<00:00, 14.33it/s]


train loss of 25th epoch is 0.3981630504131317


100%|██████████| 391/391 [00:27<00:00, 14.45it/s]


train loss of 26th epoch is 0.5129622220993042


100%|██████████| 391/391 [00:27<00:00, 14.36it/s]


train loss of 27th epoch is 0.21013765037059784


100%|██████████| 391/391 [00:27<00:00, 14.29it/s]


train loss of 28th epoch is 0.3162614107131958


100%|██████████| 391/391 [00:27<00:00, 14.28it/s]


train loss of 29th epoch is 0.3651740849018097


100%|██████████| 391/391 [00:27<00:00, 14.34it/s]


train loss of 30th epoch is 0.19125953316688538


100%|██████████| 391/391 [00:27<00:00, 14.17it/s]


train loss of 31th epoch is 0.3599313199520111


100%|██████████| 391/391 [00:27<00:00, 14.33it/s]


train loss of 32th epoch is 0.22059869766235352


100%|██████████| 391/391 [00:26<00:00, 14.50it/s]


train loss of 33th epoch is 0.47624802589416504


100%|██████████| 391/391 [00:27<00:00, 14.43it/s]


train loss of 34th epoch is 0.21259406208992004


100%|██████████| 391/391 [00:27<00:00, 14.45it/s]


train loss of 35th epoch is 0.22828619182109833


100%|██████████| 391/391 [00:27<00:00, 14.25it/s]


train loss of 36th epoch is 0.29365068674087524


100%|██████████| 391/391 [00:27<00:00, 14.31it/s]


train loss of 37th epoch is 0.37277698516845703


100%|██████████| 391/391 [00:27<00:00, 14.31it/s]


train loss of 38th epoch is 0.26651519536972046


100%|██████████| 391/391 [00:27<00:00, 14.16it/s]


train loss of 39th epoch is 0.17823240160942078


100%|██████████| 391/391 [00:28<00:00, 13.88it/s]


train loss of 40th epoch is 0.12389031797647476


100%|██████████| 391/391 [00:27<00:00, 14.23it/s]


train loss of 41th epoch is 0.353776752948761


100%|██████████| 391/391 [00:27<00:00, 14.39it/s]


train loss of 42th epoch is 0.28888192772865295


100%|██████████| 391/391 [00:27<00:00, 14.39it/s]


train loss of 43th epoch is 0.4029621481895447


100%|██████████| 391/391 [00:27<00:00, 14.41it/s]


train loss of 44th epoch is 0.29007840156555176


100%|██████████| 391/391 [00:27<00:00, 14.39it/s]


train loss of 45th epoch is 0.1798766404390335


100%|██████████| 391/391 [00:26<00:00, 14.48it/s]


train loss of 46th epoch is 0.1956147998571396


100%|██████████| 391/391 [00:27<00:00, 14.37it/s]


train loss of 47th epoch is 0.2774374186992645


100%|██████████| 391/391 [00:27<00:00, 14.23it/s]


train loss of 48th epoch is 0.38805148005485535


100%|██████████| 391/391 [00:27<00:00, 14.44it/s]


train loss of 49th epoch is 0.35837075114250183


100%|██████████| 391/391 [00:27<00:00, 14.28it/s]


train loss of 50th epoch is 0.21387335658073425


100%|██████████| 391/391 [00:27<00:00, 14.46it/s]


train loss of 51th epoch is 0.19600890576839447


100%|██████████| 391/391 [00:26<00:00, 14.52it/s]


train loss of 52th epoch is 0.43613117933273315


100%|██████████| 391/391 [00:26<00:00, 14.49it/s]


train loss of 53th epoch is 0.27982157468795776


100%|██████████| 391/391 [00:27<00:00, 14.46it/s]


train loss of 54th epoch is 0.3019636571407318


100%|██████████| 391/391 [00:26<00:00, 14.50it/s]


train loss of 55th epoch is 0.2859998047351837


100%|██████████| 391/391 [00:27<00:00, 14.44it/s]


train loss of 56th epoch is 0.21031005680561066


100%|██████████| 391/391 [00:27<00:00, 14.26it/s]


train loss of 57th epoch is 0.2518811821937561


100%|██████████| 391/391 [00:27<00:00, 14.31it/s]


train loss of 58th epoch is 0.14825479686260223


100%|██████████| 391/391 [00:27<00:00, 14.41it/s]


train loss of 59th epoch is 0.3019539713859558


100%|██████████| 391/391 [00:27<00:00, 14.39it/s]


train loss of 60th epoch is 0.38178354501724243


100%|██████████| 391/391 [00:27<00:00, 14.40it/s]


train loss of 61th epoch is 0.22477968037128448


100%|██████████| 391/391 [00:27<00:00, 14.45it/s]


train loss of 62th epoch is 0.2615518271923065


100%|██████████| 391/391 [00:27<00:00, 14.38it/s]


train loss of 63th epoch is 0.2708331346511841


100%|██████████| 391/391 [00:26<00:00, 14.58it/s]


train loss of 64th epoch is 0.20933017134666443


100%|██████████| 391/391 [00:26<00:00, 14.54it/s]


train loss of 65th epoch is 0.17214587330818176


100%|██████████| 391/391 [00:27<00:00, 14.21it/s]


train loss of 66th epoch is 0.25782617926597595


100%|██████████| 391/391 [00:26<00:00, 14.61it/s]


train loss of 67th epoch is 0.31490081548690796


100%|██████████| 391/391 [00:26<00:00, 14.50it/s]


train loss of 68th epoch is 0.384280800819397


100%|██████████| 391/391 [00:26<00:00, 14.61it/s]


train loss of 69th epoch is 0.16593846678733826


100%|██████████| 391/391 [00:26<00:00, 14.65it/s]


train loss of 70th epoch is 0.26213976740837097


100%|██████████| 391/391 [00:26<00:00, 14.58it/s]


train loss of 71th epoch is 0.192374125123024


100%|██████████| 391/391 [00:26<00:00, 14.63it/s]


train loss of 72th epoch is 0.24890784919261932


100%|██████████| 391/391 [00:26<00:00, 14.59it/s]


train loss of 73th epoch is 0.3041225075721741


100%|██████████| 391/391 [00:26<00:00, 14.61it/s]


train loss of 74th epoch is 0.31478771567344666


100%|██████████| 391/391 [00:26<00:00, 14.62it/s]


train loss of 75th epoch is 0.3192918300628662


100%|██████████| 391/391 [00:26<00:00, 14.58it/s]


train loss of 76th epoch is 0.3871925175189972


100%|██████████| 391/391 [00:26<00:00, 14.64it/s]


train loss of 77th epoch is 0.27443885803222656


100%|██████████| 391/391 [00:26<00:00, 14.60it/s]


train loss of 78th epoch is 0.20707833766937256


100%|██████████| 391/391 [00:26<00:00, 14.67it/s]


train loss of 79th epoch is 0.2929301857948303


100%|██████████| 391/391 [00:26<00:00, 14.64it/s]


train loss of 80th epoch is 0.20139849185943604


100%|██████████| 391/391 [00:26<00:00, 14.71it/s]


train loss of 81th epoch is 0.1713748574256897


100%|██████████| 391/391 [00:26<00:00, 14.69it/s]


train loss of 82th epoch is 0.07828971743583679


100%|██████████| 391/391 [00:26<00:00, 14.75it/s]


train loss of 83th epoch is 0.06522605568170547


100%|██████████| 391/391 [00:26<00:00, 14.72it/s]


train loss of 84th epoch is 0.14986944198608398


100%|██████████| 391/391 [00:26<00:00, 14.55it/s]


train loss of 85th epoch is 0.17571017146110535


100%|██████████| 391/391 [00:26<00:00, 14.70it/s]


train loss of 86th epoch is 0.05661093443632126


100%|██████████| 391/391 [00:26<00:00, 14.79it/s]


train loss of 87th epoch is 0.10613907873630524


100%|██████████| 391/391 [00:26<00:00, 14.68it/s]


train loss of 88th epoch is 0.033553335815668106


100%|██████████| 391/391 [00:26<00:00, 14.75it/s]


train loss of 89th epoch is 0.028850603848695755


100%|██████████| 391/391 [00:26<00:00, 14.58it/s]


train loss of 90th epoch is 0.12916778028011322


100%|██████████| 391/391 [00:26<00:00, 14.68it/s]


train loss of 91th epoch is 0.053967636078596115


100%|██████████| 391/391 [00:26<00:00, 14.78it/s]


train loss of 92th epoch is 0.0817098319530487


100%|██████████| 391/391 [00:26<00:00, 14.70it/s]


train loss of 93th epoch is 0.02702506259083748


100%|██████████| 391/391 [00:26<00:00, 14.94it/s]


train loss of 94th epoch is 0.04711759090423584


100%|██████████| 391/391 [00:27<00:00, 14.34it/s]


train loss of 95th epoch is 0.09532084316015244


100%|██████████| 391/391 [00:26<00:00, 14.58it/s]


train loss of 96th epoch is 0.07855679094791412


100%|██████████| 391/391 [00:26<00:00, 14.82it/s]


train loss of 97th epoch is 0.02111576497554779


100%|██████████| 391/391 [00:26<00:00, 14.81it/s]


train loss of 98th epoch is 0.06760906428098679


100%|██████████| 391/391 [00:26<00:00, 14.55it/s]


train loss of 99th epoch is 0.08011358231306076


100%|██████████| 391/391 [00:26<00:00, 14.70it/s]


train loss of 100th epoch is 0.0456092394888401


100%|██████████| 391/391 [00:26<00:00, 14.62it/s]


train loss of 101th epoch is 0.06793829053640366


100%|██████████| 391/391 [00:26<00:00, 14.80it/s]


train loss of 102th epoch is 0.042468491941690445


100%|██████████| 391/391 [00:26<00:00, 14.83it/s]


train loss of 103th epoch is 0.05282144993543625


100%|██████████| 391/391 [00:26<00:00, 14.64it/s]


train loss of 104th epoch is 0.02620803378522396


100%|██████████| 391/391 [00:26<00:00, 14.75it/s]


train loss of 105th epoch is 0.1355040967464447


100%|██████████| 391/391 [00:26<00:00, 14.93it/s]


train loss of 106th epoch is 0.07132580876350403


100%|██████████| 391/391 [00:26<00:00, 14.53it/s]


train loss of 107th epoch is 0.054990749806165695


100%|██████████| 391/391 [00:26<00:00, 14.67it/s]


train loss of 108th epoch is 0.0720004290342331


100%|██████████| 391/391 [00:26<00:00, 14.65it/s]


train loss of 109th epoch is 0.06327695399522781


100%|██████████| 391/391 [00:26<00:00, 14.67it/s]


train loss of 110th epoch is 0.05284396559000015


100%|██████████| 391/391 [00:26<00:00, 14.74it/s]


train loss of 111th epoch is 0.0853373110294342


100%|██████████| 391/391 [00:26<00:00, 14.71it/s]


train loss of 112th epoch is 0.0561394989490509


100%|██████████| 391/391 [00:26<00:00, 14.68it/s]


train loss of 113th epoch is 0.05283202975988388


100%|██████████| 391/391 [00:26<00:00, 14.66it/s]


train loss of 114th epoch is 0.054164402186870575


100%|██████████| 391/391 [00:26<00:00, 14.63it/s]


train loss of 115th epoch is 0.021254464983940125


100%|██████████| 391/391 [00:25<00:00, 15.06it/s]


train loss of 116th epoch is 0.06397504359483719


100%|██████████| 391/391 [00:26<00:00, 14.94it/s]


train loss of 117th epoch is 0.028058189898729324


100%|██████████| 391/391 [00:26<00:00, 15.00it/s]


train loss of 118th epoch is 0.02337826043367386


100%|██████████| 391/391 [00:25<00:00, 15.04it/s]


train loss of 119th epoch is 0.044148195534944534


100%|██████████| 391/391 [00:25<00:00, 15.04it/s]


train loss of 120th epoch is 0.04635099321603775


100%|██████████| 391/391 [00:26<00:00, 14.92it/s]


train loss of 121th epoch is 0.03703003749251366


100%|██████████| 391/391 [00:25<00:00, 15.09it/s]


train loss of 122th epoch is 0.047396015375852585


100%|██████████| 391/391 [00:25<00:00, 15.08it/s]


train loss of 123th epoch is 0.0649748370051384


100%|██████████| 391/391 [00:26<00:00, 14.59it/s]


train loss of 124th epoch is 0.014022232964634895


100%|██████████| 391/391 [00:25<00:00, 15.05it/s]


train loss of 125th epoch is 0.07020173966884613


100%|██████████| 391/391 [00:25<00:00, 15.06it/s]


train loss of 126th epoch is 0.007904413156211376


100%|██████████| 391/391 [00:26<00:00, 14.93it/s]


train loss of 127th epoch is 0.02732989192008972


100%|██████████| 391/391 [00:25<00:00, 15.11it/s]


train loss of 128th epoch is 0.034730926156044006


100%|██████████| 391/391 [00:25<00:00, 15.09it/s]


train loss of 129th epoch is 0.009757854044437408


100%|██████████| 391/391 [00:26<00:00, 15.01it/s]


train loss of 130th epoch is 0.010307298973202705


100%|██████████| 391/391 [00:26<00:00, 15.01it/s]


train loss of 131th epoch is 0.03910099342465401


100%|██████████| 391/391 [00:26<00:00, 15.00it/s]


train loss of 132th epoch is 0.00895288959145546


100%|██████████| 391/391 [00:26<00:00, 14.74it/s]


train loss of 133th epoch is 0.008872587233781815


100%|██████████| 391/391 [00:26<00:00, 14.76it/s]


train loss of 134th epoch is 0.009057307615876198


100%|██████████| 391/391 [00:26<00:00, 14.89it/s]


train loss of 135th epoch is 0.02063699997961521


100%|██████████| 391/391 [00:26<00:00, 14.96it/s]


train loss of 136th epoch is 0.0031861893367022276


100%|██████████| 391/391 [00:25<00:00, 15.09it/s]


train loss of 137th epoch is 0.010832625441253185


100%|██████████| 391/391 [00:26<00:00, 15.00it/s]


train loss of 138th epoch is 0.007335912436246872


100%|██████████| 391/391 [00:26<00:00, 14.87it/s]


train loss of 139th epoch is 0.023749321699142456


100%|██████████| 391/391 [00:26<00:00, 15.01it/s]


train loss of 140th epoch is 0.014112032949924469


100%|██████████| 391/391 [00:25<00:00, 15.06it/s]


train loss of 141th epoch is 0.07767067104578018


100%|██████████| 391/391 [00:26<00:00, 14.91it/s]


train loss of 142th epoch is 0.004917144309729338


100%|██████████| 391/391 [00:26<00:00, 14.73it/s]


train loss of 143th epoch is 0.03595484420657158


100%|██████████| 391/391 [00:26<00:00, 14.85it/s]


train loss of 144th epoch is 0.015748323872685432


100%|██████████| 391/391 [00:26<00:00, 15.03it/s]


train loss of 145th epoch is 0.007093379739671946


100%|██████████| 391/391 [00:26<00:00, 14.97it/s]


train loss of 146th epoch is 0.022776693105697632


100%|██████████| 391/391 [00:25<00:00, 15.10it/s]


train loss of 147th epoch is 0.02105460688471794


100%|██████████| 391/391 [00:26<00:00, 14.96it/s]


train loss of 148th epoch is 0.04173686355352402


100%|██████████| 391/391 [00:26<00:00, 14.97it/s]


train loss of 149th epoch is 0.04062512889504433


100%|██████████| 391/391 [00:25<00:00, 15.11it/s]


train loss of 150th epoch is 0.01838531717658043


100%|██████████| 391/391 [00:26<00:00, 15.01it/s]


train loss of 151th epoch is 0.010985726490616798


100%|██████████| 391/391 [00:26<00:00, 14.95it/s]


train loss of 152th epoch is 0.02045099064707756


100%|██████████| 391/391 [00:26<00:00, 14.82it/s]


train loss of 153th epoch is 0.009561299346387386


100%|██████████| 391/391 [00:26<00:00, 14.83it/s]


train loss of 154th epoch is 0.0451299212872982


100%|██████████| 391/391 [00:25<00:00, 15.05it/s]


train loss of 155th epoch is 0.0173789169639349


100%|██████████| 391/391 [00:26<00:00, 15.01it/s]


train loss of 156th epoch is 0.0736633762717247


100%|██████████| 391/391 [00:25<00:00, 15.10it/s]


train loss of 157th epoch is 0.026433875784277916


100%|██████████| 391/391 [00:25<00:00, 15.05it/s]


train loss of 158th epoch is 0.04039289802312851


100%|██████████| 391/391 [00:25<00:00, 15.06it/s]


train loss of 159th epoch is 0.04451052471995354


100%|██████████| 391/391 [00:26<00:00, 15.02it/s]


train loss of 160th epoch is 0.02748122252523899


100%|██████████| 391/391 [00:26<00:00, 15.03it/s]


train loss of 161th epoch is 0.02905499003827572


100%|██████████| 391/391 [00:26<00:00, 15.02it/s]

train loss of 162th epoch is 0.08574984967708588
Finished Training


# Test

- model.pth   

100 100 Loss: 0.343 | Acc: 91.560% (9156/10000)


- 30.pth

100 100 Loss: 0.468 | Acc: 84.730% (8473/10000)


- 60.pth


100 100 Loss: 0.438 | Acc: 86.170% (8617/10000)

- 90.pth


100 100 Loss: 0.279 | Acc: 91.710% (9171/10000)

- 120.pth

100 100 Loss: 0.353 | Acc: 91.210% (9121/10000)


- 150.pth


100 Loss: 0.342 | Acc: 91.450% (9145/10000)


In [23]:
def test():

    ####### not touch
    model_name = MyModel
    # /content/models/230412_2145/0.pth
    #/content/models/230412_2208/1.pth
    #/content/models/230416_0927/90.pth
    #/content/models/230417_0009/30.pth
    checkpoint_path = './models/230417_0009/150.pth'  #수정 필요
    mode = 'test' 
#     data_dir = "./test_data"
#     meta_path = "./answer.json"
    model = get_model(model_name,checkpoint_path)
    #######
    
#     # Create training and validation datasets
#     test_datasets = MyDataset(meta_path, data_dir, data_transforms['test'],is_train= False) #self,meta_path,root_dir,transform=None, is_train = True

#     # Create training and validation dataloaders
#     test_dataloader = torch.utils.data.DataLoader(test_datasets, batch_size=batch_size, shuffle=False, num_workers=4)

    test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
    
    testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)
    
    # Detect if we have a GPU available
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    # Send the model to GPU
    #multi-GPU
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model = model.to(device)

    # Set model as evaluation mode
    for param in model.parameters():
        param.requires_grad = False
    model.eval()
    
    # Inference
    test_loss = 0
    correct = 0
    total = 0
    criterion = nn.CrossEntropyLoss()
    with torch.no_grad():
        batch_idx = 0
        for inputs, targets in tqdm(testloader):
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            
            test_loss += loss.item()
            _, predicted = outputs.max(1)
            total += targets.size(0)
            correct += predicted.eq(targets).sum().item()
            batch_idx = batch_idx + 1
            print(batch_idx, len(testloader), 'Loss: %.3f | Acc: %.3f%% (%d/%d)'
                         % (test_loss/(batch_idx+1), 100.*correct/total, correct, total))



# Main

In [24]:

def main() :

    random.seed(53)
    test()

   
if __name__ == '__main__':
    main()

start loading model parameter
model parameter loaded
Files already downloaded and verified


  5%|▌         | 5/100 [00:00<00:05, 17.61it/s]

1 100 Loss: 0.141 | Acc: 96.000% (96/100)
2 100 Loss: 0.198 | Acc: 93.500% (187/200)
3 100 Loss: 0.236 | Acc: 92.667% (278/300)
4 100 Loss: 0.239 | Acc: 92.750% (371/400)
5 100 Loss: 0.231 | Acc: 93.600% (468/500)
6 100 Loss: 0.218 | Acc: 94.167% (565/600)


 12%|█▏        | 12/100 [00:00<00:03, 26.75it/s]

7 100 Loss: 0.223 | Acc: 94.143% (659/700)
8 100 Loss: 0.248 | Acc: 93.500% (748/800)
9 100 Loss: 0.273 | Acc: 93.000% (837/900)
10 100 Loss: 0.282 | Acc: 92.600% (926/1000)
11 100 Loss: 0.280 | Acc: 92.455% (1017/1100)
12 100 Loss: 0.295 | Acc: 92.000% (1104/1200)
13 100 Loss: 0.294 | Acc: 91.769% (1193/1300)


 18%|█▊        | 18/100 [00:00<00:03, 26.43it/s]

14 100 Loss: 0.295 | Acc: 91.714% (1284/1400)
15 100 Loss: 0.288 | Acc: 92.000% (1380/1500)
16 100 Loss: 0.293 | Acc: 91.938% (1471/1600)
17 100 Loss: 0.294 | Acc: 92.059% (1565/1700)
18 100 Loss: 0.298 | Acc: 91.889% (1654/1800)
19 100 Loss: 0.304 | Acc: 91.684% (1742/1900)


 26%|██▌       | 26/100 [00:01<00:02, 29.64it/s]

20 100 Loss: 0.313 | Acc: 91.500% (1830/2000)
21 100 Loss: 0.320 | Acc: 91.381% (1919/2100)
22 100 Loss: 0.316 | Acc: 91.500% (2013/2200)
23 100 Loss: 0.332 | Acc: 91.130% (2096/2300)
24 100 Loss: 0.332 | Acc: 91.083% (2186/2400)
25 100 Loss: 0.338 | Acc: 91.040% (2276/2500)
26 100 Loss: 0.354 | Acc: 91.000% (2366/2600)


 30%|███       | 30/100 [00:01<00:02, 30.76it/s]

27 100 Loss: 0.346 | Acc: 91.074% (2459/2700)
28 100 Loss: 0.343 | Acc: 91.214% (2554/2800)
29 100 Loss: 0.346 | Acc: 91.276% (2647/2900)
30 100 Loss: 0.342 | Acc: 91.333% (2740/3000)
31 100 Loss: 0.347 | Acc: 91.290% (2830/3100)
32 100 Loss: 0.347 | Acc: 91.312% (2922/3200)


 38%|███▊      | 38/100 [00:01<00:02, 28.81it/s]

33 100 Loss: 0.349 | Acc: 91.182% (3009/3300)
34 100 Loss: 0.346 | Acc: 91.147% (3099/3400)
35 100 Loss: 0.346 | Acc: 91.143% (3190/3500)
36 100 Loss: 0.347 | Acc: 91.194% (3283/3600)
37 100 Loss: 0.349 | Acc: 91.243% (3376/3700)
38 100 Loss: 0.353 | Acc: 91.158% (3464/3800)


 42%|████▏     | 42/100 [00:01<00:01, 30.31it/s]

39 100 Loss: 0.351 | Acc: 91.231% (3558/3900)
40 100 Loss: 0.352 | Acc: 91.225% (3649/4000)
41 100 Loss: 0.352 | Acc: 91.220% (3740/4100)
42 100 Loss: 0.351 | Acc: 91.238% (3832/4200)
43 100 Loss: 0.347 | Acc: 91.349% (3928/4300)
44 100 Loss: 0.348 | Acc: 91.386% (4021/4400)
45 100 Loss: 0.349 | Acc: 91.356% (4111/4500)


 50%|█████     | 50/100 [00:01<00:01, 30.26it/s]

46 100 Loss: 0.351 | Acc: 91.283% (4199/4600)
47 100 Loss: 0.351 | Acc: 91.319% (4292/4700)
48 100 Loss: 0.357 | Acc: 91.250% (4380/4800)
49 100 Loss: 0.356 | Acc: 91.306% (4474/4900)
50 100 Loss: 0.360 | Acc: 91.220% (4561/5000)
51 100 Loss: 0.357 | Acc: 91.255% (4654/5100)


 57%|█████▋    | 57/100 [00:02<00:01, 29.28it/s]

52 100 Loss: 0.354 | Acc: 91.308% (4748/5200)
53 100 Loss: 0.355 | Acc: 91.340% (4841/5300)
54 100 Loss: 0.354 | Acc: 91.352% (4933/5400)
55 100 Loss: 0.356 | Acc: 91.327% (5023/5500)
56 100 Loss: 0.359 | Acc: 91.286% (5112/5600)
57 100 Loss: 0.361 | Acc: 91.228% (5200/5700)
58 100 

 65%|██████▌   | 65/100 [00:02<00:01, 31.76it/s]

Loss: 0.361 | Acc: 91.224% (5291/5800)
59 100 Loss: 0.363 | Acc: 91.220% (5382/5900)
60 100 Loss: 0.361 | Acc: 91.167% (5470/6000)
61 100 Loss: 0.358 | Acc: 91.230% (5565/6100)
62 100 Loss: 0.359 | Acc: 91.210% (5655/6200)
63 100 Loss: 0.356 | Acc: 91.270% (5750/6300)
64 100 Loss: 0.354 | Acc: 91.312% (5844/6400)
65 100 Loss: 0.355 | Acc: 91.277% (5933/6500)


 69%|██████▉   | 69/100 [00:02<00:00, 31.20it/s]

66 100 Loss: 0.356 | Acc: 91.258% (6023/6600)
67 100 Loss: 0.352 | Acc: 91.299% (6117/6700)
68 100 Loss: 0.353 | Acc: 91.324% (6210/6800)
69 100 Loss: 0.351 | Acc: 91.348% (6303/6900)
70 100 Loss: 0.349 | Acc: 91.357% (6395/7000)
71 100 Loss: 0.352 | Acc: 91.338% (6485/7100)
72 

 77%|███████▋  | 77/100 [00:02<00:00, 31.82it/s]

100 Loss: 0.354 | Acc: 91.306% (6574/7200)
73 100 Loss: 0.354 | Acc: 91.329% (6667/7300)
74 100 Loss: 0.354 | Acc: 91.378% (6762/7400)
75 100 Loss: 0.353 | Acc: 91.387% (6854/7500)
76 100 Loss: 0.353 | Acc: 91.408% (6947/7600)
77 100 Loss: 0.355 | Acc: 91.364% (7035/7700)
78 100 Loss: 0.354 | Acc: 91.372% (7127/7800)


 85%|████████▌ | 85/100 [00:02<00:00, 31.64it/s]

79 100 Loss: 0.352 | Acc: 91.354% (7217/7900)
80 100 Loss: 0.352 | Acc: 91.350% (7308/8000)
81 100 Loss: 0.351 | Acc: 91.346% (7399/8100)
82 100 Loss: 0.350 | Acc: 91.341% (7490/8200)
83 100 Loss: 0.349 | Acc: 91.361% (7583/8300)
84 100 Loss: 0.349 | Acc: 91.321% (7671/8400)
85 100 Loss: 0.348 | Acc: 91.318% (7762/8500)


 89%|████████▉ | 89/100 [00:03<00:00, 31.91it/s]

86 100 Loss: 0.353 | Acc: 91.256% (7848/8600)
87 100 Loss: 0.351 | Acc: 91.287% (7942/8700)
88 100 Loss: 0.352 | Acc: 91.273% (8032/8800)
89 100 Loss: 0.352 | Acc: 91.270% (8123/8900)
90 100 Loss: 0.351 | Acc: 91.289% (8216/9000)
91 100 Loss: 0.349 | Acc: 91.330% (8311/9100)
92 100 Loss: 0.349 | Acc: 91.315% (8401/9200)


 97%|█████████▋| 97/100 [00:03<00:00, 31.17it/s]

93 100 Loss: 0.349 | Acc: 91.301% (8491/9300)
94 100 Loss: 0.348 | Acc: 91.319% (8584/9400)
95 100 Loss: 0.345 | Acc: 91.358% (8679/9500)
96 100 Loss: 0.345 | Acc: 91.365% (8771/9600)
97 100 Loss: 0.343 | Acc: 91.423% (8868/9700)
98 100 Loss: 0.342 | Acc: 91.459% (8963/9800)
99 100 Loss: 0.343 | Acc: 91.444% (9053/9900)
100

100%|██████████| 100/100 [00:03<00:00, 28.33it/s]

 100 Loss: 0.342 | Acc: 91.450% (9145/10000)


In [ ]:
def test():

    ####### not touch
    model_name = MyModel
    # /content/models/230412_2145/0.pth
    #/content/models/230412_2208/1.pth
    #/content/models/230415_0435/160.pth
    checkpoint_path = './models/230415_0435/150.pth'  #수정 필요
    mode = 'test' 
#     data_dir = "./test_data"
#     meta_path = "./answer.json"
    model = get_model(model_name,checkpoint_path)
    #######
    
#     # Create training and validation datasets
#     test_datasets = MyDataset(meta_path, data_dir, data_transforms['test'],is_train= False) #self,meta_path,root_dir,transform=None, is_train = True

#     # Create training and validation dataloaders
#     test_dataloader = torch.utils.data.DataLoader(test_datasets, batch_size=batch_size, shuffle=False, num_workers=4)

    test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
    
    testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)
    
    # Detect if we have a GPU available
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    # Send the model to GPU
    #multi-GPU
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model = model.to(device)

    # Set model as evaluation mode
    for param in model.parameters():
        param.requires_grad = False
    model.eval()
    
    # Inference
    test_loss = 0
    correct = 0
    total = 0
    criterion = nn.CrossEntropyLoss()
    with torch.no_grad():
        batch_idx = 0
        for inputs, targets in tqdm(testloader):
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            
            test_loss += loss.item()
            _, predicted = outputs.max(1)
            total += targets.size(0)
            correct += predicted.eq(targets).sum().item()
            batch_idx = batch_idx + 1
            print(batch_idx, len(testloader), 'Loss: %.3f | Acc: %.3f%% (%d/%d)'
                         % (test_loss/(batch_idx+1), 100.*correct/total, correct, total))



In [ ]:

def main() :

    random.seed(53)
    test()

   
if __name__ == '__main__':
    main()

start loading model parameter
model parameter loaded
Files already downloaded and verified


  6%|▌         | 6/100 [00:00<00:03, 25.81it/s]

1 100 Loss: 0.493 | Acc: 75.000% (75/100)
2 100 Loss: 0.714 | Acc: 71.500% (143/200)
3 100 Loss: 0.808 | Acc: 71.333% (214/300)
4 100 Loss: 0.836 | Acc: 72.250% (289/400)
5 100 Loss: 0.862 | Acc: 71.600% (358/500)
6 100 Loss: 0.835 | Acc: 71.667% (430/600)
7 100 Loss: 0.871 | Acc: 71.143% (498/700)
8 100 Loss: 0.907 | Acc: 71.000% (568/800)
9 100 Loss: 0.948 | Acc: 70.333% (633/900)
10 100 Loss: 0.974 | Acc: 70.400% (704/1000)


 17%|█▋        | 17/100 [00:00<00:01, 41.56it/s]

11 100 Loss: 0.965 | Acc: 70.909% (780/1100)
12 100 Loss: 0.986 | Acc: 70.667% (848/1200)
13 100 Loss: 0.981 | Acc: 70.846% (921/1300)
14 100 Loss: 0.994 | Acc: 70.000% (980/1400)
15 100 Loss: 1.018 | Acc: 69.533% (1043/1500)
16 100 Loss: 1.034 | Acc: 69.000% (1104/1600)
17 100 Loss: 1.038 | Acc: 69.353% (1179/1700)
18 100 Loss: 1.056 | Acc: 69.000% (1242/1800)
19 100 Loss: 1.061 | Acc: 68.684% (1305/1900)
20 100 Loss: 1.076 | Acc: 68.550% (1371/2000)
21 100 Loss: 1.081 | Acc: 68.143% (1431/2100)


 28%|██▊       | 28/100 [00:00<00:01, 45.39it/s]

22 100 Loss: 1.099 | Acc: 67.909% (1494/2200)
23 100 Loss: 1.095 | Acc: 68.087% (1566/2300)
24 100 Loss: 1.101 | Acc: 68.042% (1633/2400)
25 100 Loss: 1.106 | Acc: 67.960% (1699/2500)
26 100 Loss: 1.120 | Acc: 67.731% (1761/2600)
27 100 Loss: 1.126 | Acc: 67.556% (1824/2700)
28 100 Loss: 1.126 | Acc: 67.643% (1894/2800)
29 100 Loss: 1.132 | Acc: 67.655% (1962/2900)
30 100 Loss: 1.133 | Acc: 67.533% (2026/3000)
31 100 Loss: 1.132 | Acc: 67.548% (2094/3100)
32 100 Loss: 1.129 | Acc: 67.625% (2164/3200)


 40%|████      | 40/100 [00:00<00:01, 48.83it/s]

33 100 Loss: 1.130 | Acc: 67.697% (2234/3300)
34 100 Loss: 1.135 | Acc: 67.529% (2296/3400)
35 100 Loss: 1.136 | Acc: 67.600% (2366/3500)
36 100 Loss: 1.136 | Acc: 67.583% (2433/3600)
37 100 Loss: 1.134 | Acc: 67.622% (2502/3700)
38 100 Loss: 1.139 | Acc: 67.447% (2563/3800)
39 100 Loss: 1.131 | Acc: 67.513% (2633/3900)
40 100 Loss: 1.125 | Acc: 67.625% (2705/4000)
41 100 Loss: 1.122 | Acc: 67.683% (2775/4100)
42 100 Loss: 1.124 | Acc: 67.619% (2840/4200)


 51%|█████     | 51/100 [00:01<00:00, 49.66it/s]

43 100 Loss: 1.131 | Acc: 67.674% (2910/4300)
44 100 Loss: 1.126 | Acc: 67.864% (2986/4400)
45 100 Loss: 1.123 | Acc: 67.956% (3058/4500)
46 100 Loss: 1.124 | Acc: 67.913% (3124/4600)
47 100 Loss: 1.118 | Acc: 67.979% (3195/4700)
48 100 Loss: 1.119 | Acc: 67.896% (3259/4800)
49 100 Loss: 1.116 | Acc: 68.020% (3333/4900)
50 100 Loss: 1.112 | Acc: 68.100% (3405/5000)
51 100 Loss: 1.114 | Acc: 68.039% (3470/5100)
52 100 Loss: 1.116 | Acc: 68.038% (3538/5200)
53 100 Loss: 1.115 | Acc: 67.981% (3603/5300)


 61%|██████    | 61/100 [00:01<00:00, 48.96it/s]

54 100 Loss: 1.110 | Acc: 68.019% (3673/5400)
55 100 Loss: 1.106 | Acc: 68.145% (3748/5500)
56 100 Loss: 1.111 | Acc: 68.036% (3810/5600)
57 100 Loss: 1.110 | Acc: 68.053% (3879/5700)
58 100 Loss: 1.105 | Acc: 68.121% (3951/5800)
59 100 Loss: 1.105 | Acc: 68.068% (4016/5900)
60 100 Loss: 1.109 | Acc: 67.833% (4070/6000)
61 100 Loss: 1.108 | Acc: 67.885% (4141/6100)
62 100 Loss: 1.108 | Acc: 67.855% (4207/6200)
63 100 Loss: 1.106 | Acc: 67.889% (4277/6300)
64 100 Loss: 1.105 | Acc: 67.953% (4349/6400)


 73%|███████▎  | 73/100 [00:01<00:00, 52.57it/s]

65 100 Loss: 1.105 | Acc: 67.954% (4417/6500)
66 100 Loss: 1.105 | Acc: 67.955% (4485/6600)
67 100 Loss: 1.104 | Acc: 67.925% (4551/6700)
68 100 Loss: 1.102 | Acc: 68.000% (4624/6800)
69 100 Loss: 1.101 | Acc: 68.029% (4694/6900)
70 100 Loss: 1.101 | Acc: 68.000% (4760/7000)
71 100 Loss: 1.104 | Acc: 67.944% (4824/7100)
72 100 Loss: 1.108 | Acc: 67.944% (4892/7200)
73 100 Loss: 1.105 | Acc: 67.945% (4960/7300)
74 100 Loss: 1.102 | Acc: 67.986% (5031/7400)
75 100 Loss: 1.105 | Acc: 67.933% (5095/7500)
76 100 Loss: 1.105 | Acc: 67.908% (5161/7600)


 85%|████████▌ | 85/100 [00:01<00:00, 50.68it/s]

77 100 Loss: 1.106 | Acc: 67.870% (5226/7700)
78 100 Loss: 1.108 | Acc: 67.910% (5297/7800)
79 100 Loss: 1.108 | Acc: 67.975% (5370/7900)
80 100 Loss: 1.110 | Acc: 67.963% (5437/8000)
81 100 Loss: 1.108 | Acc: 67.963% (5505/8100)
82 100 Loss: 1.105 | Acc: 67.976% (5574/8200)
83 100 Loss: 1.102 | Acc: 68.060% (5649/8300)
84 100 Loss: 1.104 | Acc: 68.024% (5714/8400)
85 100 Loss: 1.107 | Acc: 67.988% (5779/8500)
86 100 Loss: 1.109 | Acc: 67.942% (5843/8600)
87

 97%|█████████▋| 97/100 [00:02<00:00, 51.78it/s]

 100 Loss: 1.108 | Acc: 67.989% (5915/8700)
88 100 Loss: 1.106 | Acc: 68.057% (5989/8800)
89 100 Loss: 1.107 | Acc: 67.989% (6051/8900)
90 100 Loss: 1.106 | Acc: 67.956% (6116/9000)
91 100 Loss: 1.105 | Acc: 67.901% (6179/9100)
92 100 Loss: 1.104 | Acc: 67.870% (6244/9200)
93 100 Loss: 1.107 | Acc: 67.806% (6306/9300)
94 100 Loss: 1.107 | Acc: 67.777% (6371/9400)
95 100 Loss: 1.107 | Acc: 67.811% (6442/9500)
96 100 Loss: 1.105 | Acc: 67.844% (6513/9600)
97 100 Loss: 1.103 | Acc: 67.876% (6584/9700)


100%|██████████| 100/100 [00:02<00:00, 45.97it/s]

98 100 Loss: 1.105 | Acc: 67.857% (6650/9800)
99 100 Loss: 1.105 | Acc: 67.828% (6715/9900)
100 100 Loss: 1.105 | Acc: 67.850% (6785/10000)
